### PR TITLE
RHS Subscriptions

### PR DESCRIPTION
**Pull request by lieut-data**
#### Summary
Highlight channel subscriptions in the RHS:

image

When there are no subscriptions:

image

And when not signed in:

image

For this change, I tried to stick with the existing dependencies as I didn't want to take on ownership of upgrading `package-lock.json`. (The styles would have been much clearer with `styled-components`, or even `SASS`.)

For context on this feature, see the private channel https://community.mattermost.com/private-core/channels/gitlab-strategic-integration.

### CODE REVIEW

**Comment by lieut-data** [comment]
The tests are failing, it seems because there simply is no `test` target for the webapp. Wondering if something changed in the upstream orbs?

File: server/plugin.go
Comment (lieut-data) [inline server/plugin.go:603]: This constrains the websocket event to members of the channel, /but/ I see there is logic in the plugin around some users being denied access to some subset of projects. Worse, it looks like these permission checks actually involve a round trip per user -- is there some background information I can learn about?

File: server/api.go
Comment (ashishbhate) [inline server/api.go:650]: We should probably check the err returned here. IIRC, we've had some incidents on the server around JSON Marshalling so a quick error check helps.

File: server/plugin.go
Comment (ashishbhate) [reply server/plugin.go:588]: IIUC, but isn't it implied that access to the channel gives access to the subscription information? 
So shouldn't this be left to channel admins (by controlling channel memberships)?

**Review by ashishbhate — APPROVED** [review]
LGTM, I don't know the Gitlab side of things very well (or plugins in general, tbh) so my review focuses on the actual written code. 

I'll defer to the other reviewers on the high-level view.

File: server/plugin.go
Comment (mickmister) [reply server/plugin.go:588]: @lieut-data It looks like the `/gitlab subscriptions list` command does not filter on any sort of permission checks, which is sort of the equivalent of this websocket message right? I'm thinking we may not need any further access control other than scoping it to the specific channel, as you've done here.

**Review by mickmister — COMMENTED** [review]
Great work @lieut-data! I left some non-blocking suggestions and questions. Let me know what you think :+1:

File: server/api.go
Comment (ashishbhate) [inline server/api.go:623]: I just ran into this while deploying:

`JoinPath` was added in Go 1.19 https://pkg.go.dev/net/url#URL.JoinPath

The go.mod for this project states that it works with go 1.16.

The mattermost-server project requires go 1.18.

File: .nvmrc
Comment (lieut-data) [reply .nvmrc:1]: Ah, good catch! When my first attempt to compile and deploy failed, I assumed it was the node version and mistakenly copied the version from webapp, then opted not to `git add package-lock.json` but forgot about my changes here in `.nvmrc`. I'll fix this to point at the latest v14 here which I believe uses the old lock version.

File: webapp/src/components/rhs_sidebar/index.js
Comment (lieut-data) [reply webapp/src/components/rhs_sidebar/index.js:25]: Ooh, thank you for reminding me of this! Fixing with a `const noSubscriptions`.

File: webapp/src/actions/index.js
Comment (lieut-data) [reply webapp/src/actions/index.js:238]: I'll copy the pattern above and return `{subscriptions}` for consistency -- thank you, and no worries re: TypeScript.

File: webapp/src/components/rhs_sidebar/rhs_sidebar.jsx
Comment (lieut-data) [reply webapp/src/components/rhs_sidebar/rhs_sidebar.jsx:267]: Yes, definitely!

File: webapp/src/components/rhs_sidebar/rhs_sidebar.jsx
Comment (lieut-data) [reply webapp/src/components/rhs_sidebar/rhs_sidebar.jsx:331]: Good point, if I switch between channels I should ensure this reconciles correctly.

File: server/plugin.go
Comment (lieut-data) [reply server/plugin.go:588]: Awesome, thank you for this feedback -- will leave this as-is.

File: server/api.go
Comment (lieut-data) [reply server/api.go:623]: Thank you for this: I didn't even check the supported versions on this method!

File: webapp/src/components/rhs_sidebar/rhs_sidebar.jsx
Comment (mickmister) [reply webapp/src/components/rhs_sidebar/rhs_sidebar.jsx]: @lieut-data In general, our plugins have been using `scss`. I know Playbooks is using `styled-components` and no `scss` now, though I'd like to keep consistency with the other plugin projects our team manages in this aspect.

Using `css` instead of `scss` can be a pain, because you don't have the same easily accessible nesting capabilities with `css`. Of course there's the equivalent syntax in `css`, which may not be too bad in this case.

I'll leave it up to you to decide how to apply styles here. I'm 2/5 on not introducing `styled-components` here though. Hopefully you can understand why 🙂

File: .nvmrc
Comment (mickmister) [reply .nvmrc]: Can you please ensure this works locally with `nvm install` or `nvm use` if you already have this version installed?

File: webapp/src/components/rhs_sidebar/rhs_sidebar.jsx
Comment (lieut-data) [reply webapp/src/components/rhs_sidebar/rhs_sidebar.jsx]: I'm happy to keep things simpler for the maintainers and will switch to CSS.

File: .nvmrc
Comment (mickmister) [reply .nvmrc]: If you do test this out, please remove the `node_modules` folder before testing. Likely, there won't be any issues since CI is passing

File: webapp/src/components/rhs_sidebar/rhs_sidebar.css
Comment (mickmister) [inline webapp/src/components/rhs_sidebar/rhs_sidebar.css:38]: Just so I understand, is the selector repeated to increase specificity?

File: .nvmrc
Comment (lieut-data) [reply .nvmrc]: Gah, I forgot to revert this. One second.

File: webapp/package.json
Comment (mickmister) [inline webapp/package.json:20]: We can remove this since `styled-components` is not being used anymore

**Review by mickmister — APPROVED** [review]
LGTM! Just one request to remove the `styled-components` dependency. Great work @lieut-data!

File: .nvmrc
Comment (lieut-data) [reply .nvmrc]: Wait, sorry, I've got too much on the go...

Yes, I can confirm this `.nvmrc` works locally. What I did forget to revert was the addition of `styled-components` to package.json.

Comment (lieut-data) [reply .nvmrc]: (Also triple checked after blowing away `node_modules`).

**Comment by lieut-data** [comment]
@spirosoik & @DHaussermann, is there anything you'd like to take a peek at before we merge?

**Comment by DHaussermann** [comment]
Hi @lieut-data I took a look at this PR. This is woking as expected. I did see a coupe issues.  Points 1. and 2. maybe worth investigating before we merge to see if there are quick solutions.

- Connection state and state where the signed in user is shown seem to work fine
- RHS is correctly aware of current channel open when fetching list and updated in real time as channels change
- New elements have reasonable contrast in all themes
- Performed CRUD testing on subscription 
- Adds, Edit and Deletes do all work to update the RHS view


1. The App Bar icon works as expected. But, there is no fail-over icon shown in the channel header if App Bar is in the off state (Also likely no legacy monochromatic icon exists for GitLab yet)
2. When deleting a subscription a "refresh" event such as a channel change or RHS open / close must occur for the newly deleted subscription to no longer show. (Using the `add` command to change the feature list of an existing subscription **does** update in real time       
3. When App Bar is first enabled a webapp refresh is required to see the icon. No issue if the App bar is already on when the plugin is enabled. 0/5 This feels like an App Bar issue not specific to GitLab a vaguely remember seeing this before. @mickmister may know about what causes this.

**Review by spirosoik — APPROVED** [review]
LGTM

**Comment by lieut-data** [comment]
Thanks again, @DHaussermann!

re: 1, I'm syncing with Neil on next steps: stay tuned!

re: 2, I'm not currently able to reproduce, with subscriptions being added and deleted as the slash commands are being invoked. Is there any extra detail you can help provide on how you reproduced?

https://user-images.githubusercontent.com/1023171/204812572-283df5e1-31b2-419a-9e7d-27b85ca6433e.mp4

re: 3, I agree with your assessment.

**Comment by lieut-data** [comment]
re: 1, I synced with Neil and we agreed to keep this "apps bar only". Thanks for raising that, @DHaussermann :)

**Comment by DHaussermann** [comment]
@lieut-data for 1. this is a bit strange I saw this 2 or 3 times in a row so, I thought it was consistently reproducible. Now I can nor see the issue either. Delete works in real time. I wonder if something else may have been happening on the server at the time? I've also gone back and checked 2 other servers and they are working normally as well.       

Give the relatively low impact and that I have no concrete details. I feel we should consider it resolved. If I see this again - now that I know it intermittent - I'll see if there is info on the network tab I can capture and the I can open a separate issue to prioritize and address.

**Review by DHaussermann — APPROVED** [review]
Tested and passed as per testing notes above ☝️ 

For the issues found:
1. No change will be made and the icon will only show in the App Bar
2. Is no longer reproducible. Dylan will open a new issue with console details if it is seen again
3. Is not specific to the GitHub integration     

As such no further code changes or functional testing needed for this PR.
LGTM!

@lieut-data Please merge when you have a chance.

**Comment by lieut-data** [comment]
Sweet, thank you @DHaussermann :)

**Pull request merged**
